### PR TITLE
Add big number component

### PR DESCRIPTION
**Pull request by owenatgov**
## What
Adds the big number component to the component library

## Why
We have a couple of fragmented instances of this pattern across govuk which meets our criteria for componentisation.

You can read more on the reasoning, along with a spec for the API and the markup in the comments, here: https://github.com/alphagov/govuk_publishing_components/issues/2221

## Visual Changes (from bespoke examples on production)
| Link | Before | After |
| --- | --- | --- |
| https://www.gov.uk/ | Screenshot 2021-08-20 at 16 23 15 | Screenshot 2021-08-20 at 16 23 26 |

### CODE REVIEW

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 1]
```diff
@@ -0,0 +1,75 @@
+name: Big number
+description: A big number, with a label if needed, for visualising quantitative values such as the number of government departments or historical prime ministers.
```
Comment (maxf) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:2]: It looks like that the component is only meant to display positive integers. So either you should specify that constraint and create tests for when something else is passed, or you should allow more types of numbers (looking at the code it might actually work if you pass any string, in fact)

File: app/views/govuk_publishing_components/components/big_number/_value.html.erb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    classes << "gem-c-big-number__value--decorated"
+  end
+%>
+
+<%= tag.span class: classes do %>
+  <%= number %><%= "+" if show_plus %>
+<% end %>
+
+<% unless label.nil? %>
+  <% # add a virtual space here to handle screen readers printing dictations without a space between the number and the label %>
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/big_number/_value.html.erb:18]: This is obviously an important issue - have you considered adding a test for the presence of this element in this situation? Might make it harder to accidentally remove in future.

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+body: |
+  This component needs at least one item in the `items` array passed to it, otherwise it won't render. If no `number` attribute is passed in the item, it will default to "0".
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:4]: This isn't strictly true... if I wanted to display a number 0 and nothing else, I have to pass 0 as the number attribute, otherwise the component considers the data empty and doesn't render anything.

Maybe this is trying to be too clever... why not just display whatever is passed, and if nothing is passed, don't display anything?

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+accessibility_criteria: |
+  - uses clear language and is positioned in a logical context on the page it's being used on
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:6]: Just a small readability nitpick - I think this list needs to start with something like `The component must`.

File: app/views/govuk_publishing_components/components/_big_number.html.erb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<%
+  items ||= []
+%>
+<% if items.any? %>
+  <div class="gem-c-big-number">
+    <% if items.length > 1 %>
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_big_number.html.erb:6]: Also the component sets the spacing between more than one number. What if that doesn't match the required design? Then we need an option for the layout (horizontal, vertical) and options for the space between them. Only rendering one number seems simpler.

File: app/views/govuk_publishing_components/components/big_number/_value.html.erb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  href ||= nil
+  classes = ["gem-c-big-number__value"]
+  
+  if label.nil? && href
+    classes << "gem-c-big-number__value--decorated"
+  end
+%>
+
+<%= tag.span class: classes do %>
+  <%= number %><%= "+" if show_plus %>
```
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/big_number/_value.html.erb:14]: Just checked - it renders any string passed, so this seems unnecessary.

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    description: |
+      If more than one number is present in the `items` array, it will be rendered as an unordered list as opposed to just the number element alone.
+    data:
+      items:
+        - number: 119
+        - number: 553
+  with_labels:
+    description: |
+      Labels for numbers, if present, are styled in such a way that they are inline but appear to have a block relationship. This is to adhere to established design whilst bypassing a minor issue for NVDA users where the software will identify block level items and dictate them separately. This this case it is less jarring for screen reader users to hear the number and the label as a single dictation.
+    data:
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:25]: This description is a bit wordy. Suggestion:

```
Labels for numbers are given inline styling but stacked using pseudo elements with display block. This is to bypass an issue with NVDA where block level elements are dictated separately.
```

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      items:
+        - number: 23
+          label: Ministerial departments
+        - number: 400
+          show_plus: true
+          label: Other agencies and public bodies
+  with_link:
+    description: |
+      If a `href` attribute is present on a given number, we wrap that item in an `a` tag. Only the label is underlined and underline hover styles are only applied to that label.
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:43]: This description might be unnecessary - I think this information is conveyed in the preview beneath.

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+        - number: 23
+          label: Ministerial departments
+          href: "/government/organisations#ministerial_departments"
+        - number: 400
+          show_plus: true
+          label: Other agencies and public bodies
+          href: "/government/organisations#agencies_and_other_public_bodies"
+  with_link_but_no_label:
+    description: |
+      In the event that a `href` is provided but a `label` isn't, a decorator class is applied to the number itself that mirrors the `govuk-link--no-underline` class.
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:55]: I think this could be written more clearly...

```
Numbers that are links are underlined unless a label is provided, in which case the label is given the underline.
```

**Review by andysellick — COMMENTED** [review]
Looks good, although I've raised a few thoughts that I think need to be considered.

Also I'd suggest merging these commits into a single one - as individually they don't do anything.

File: app/views/govuk_publishing_components/components/big_number/_item.html.erb [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+  href ||= nil
+%>
+
+<div class="gem-c-big-number__item">
+  <% unless href.nil? %>
+    <%= link_to href, class: "govuk-link gem-c-big-number__link" do %>
+      <%= render "govuk_publishing_components/components/big_number/value", local_assigns %>
+    <% end %>
+  <% else %>
+      <%= render "govuk_publishing_components/components/big_number/value", local_assigns %>
```
[the remark this answers is not in the document]
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/big_number/_item.html.erb:11]: That seems reasonable to have it in one place. Do you have an example of a capture in use in another component i could study?

File: app/views/govuk_publishing_components/components/_big_number.html.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/_big_number.html.erb:6]: So my thoughts on this are:

- On govuk presently, we use this component as a stacked list more often than we don't. [Example 1](https://www.gov.uk/government/how-government-works). [Example 2](https://www.gov.uk/government/get-involved). [Example 3](https://www.gov.uk/). I feel that it creates risk by not enforcing a list structure for multiple numbers, creating an (albeit quite minor) accessibility risk and effectively outsourcing the messiness across all our apps. This, whilst involving some extra calculation in the component, keeps the apps simple which is where the risk of things becoming messier is higher, in my opinion.
- I feel that the spacing is adequate both for between numbers and for between it and proceeding elements. I reckon that your suggestion for a custom margin is probably a good idea, which also is a case for only having this be one number that devs call as and when, but also I'm keen not to build in features that aren't necessary _right now_.

I'm on the fence about making this a single component in all honestly. Interested to know your thoughts on the above.

File: app/views/govuk_publishing_components/components/big_number/_value.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/big_number/_value.html.erb:14]: Max F pointed something to this effect out in his comment above. I was thinking about adding some sort of validation on the number input to dissuade wayward devs from passing words to the component, but I'm also wondering if this is overcomplicating things a bit.

File: app/views/govuk_publishing_components/components/docs/big_number.yml (hunk 3 above)
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/docs/big_number.yml:4]: I probably should've been clearer here. The 2nd sentence is for if, for some reason, other attributes like `label` are passed by the `number` attribute isn't. I suppose I could just check for the number and render nothing if it's not there?

File: app/views/govuk_publishing_components/components/big_number/_item.html.erb (hunk 10 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/big_number/_item.html.erb:11]: https://github.com/alphagov/govuk_publishing_components/blob/master/app/views/govuk_publishing_components/components/_button.html.erb#L8 is a good example, but there are plenty of others if you search.

File: app/views/govuk_publishing_components/components/_big_number.html.erb (hunk 5 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/_big_number.html.erb:6]: I didn't realise we had that many instances of multiple numbers, but I still think it should be a single thing. I think we can trust developers to do the right thing (and the list styles in question are already included in every application).

Absolutely, we shouldn't build in features that aren't needed right now, I'm thinking ahead to the complexity of unpicking this if e.g. different spacing is required, or an ordered list is required instead of an unordered one. Having it as a single thing makes it much more flexible, and reduces its complexity significantly.

File: app/views/govuk_publishing_components/components/big_number/_value.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (andysellick) [reply app/views/govuk_publishing_components/components/big_number/_value.html.erb:14]: I think that would be overcomplicating it. I think it's far easier to let developers pass any string without having to worry about what constitutes valid input - the intent of the component is clear.

(other things we might want to display that would further complicate matters should we decide to validate input include e.g. `90%`, `-20`, `1.5`, `1M`, `1Mb` - let's not go down that road)

File: app/views/govuk_publishing_components/components/docs/big_number.yml (hunk 3 above)
Comment (andysellick) [reply app/views/govuk_publishing_components/components/docs/big_number.yml:4]: Sounds good.

File: app/views/govuk_publishing_components/components/_big_number.html.erb [hunk 11]
```diff
@@ -0,0 +1,19 @@
+<%
+  items ||= []
+%>
+<% if items.any? %>
+  <div class="gem-c-big-number">
```
[the remark this answers is not in the document]
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/_big_number.html.erb:5]: Good call. I haven't been able to test this everywhere ass govuk docker has been a mess for me recently but I'll investigate spacing on this when I can.

File: app/views/govuk_publishing_components/components/big_number/_value.html.erb (hunk 6 above)
[the remark this answers is not in the document]
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/big_number/_value.html.erb:14]: THat's a very good point. I'll remove.

**Comment by owenatgov** [comment]
@andysellick I think this is ready for a re-review now. I've made a couple of updates to the spec, some of it outside your review, which you can read about here https://github.com/alphagov/govuk_publishing_components/issues/2221#issuecomment-916197999

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 12]
```diff
@@ -0,0 +1,52 @@
+name: Big number
+description: A big number, with a label if needed, for visualising quantitative values such as the number of government departments or historical prime ministers.
+body: |
+  This component requires at least a non-falsey `number` attribute value passed to it, otherwise it will not render.
+accessibility_criteria: |
+  This component must:
+  - uses clear language and is positioned in a logical context on the page it's being used on
+  - communicates number value and label (if present) in a single dictation when reading with screen reader software
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:8]: Nitpick on language here. It currently reads a little strangely.

```
This component must uses clear language and is positioned in a logical context
This component must communicates number value and label
```

Also I think accessibility criteria should define the way the component itself behaves, not how it should be used - effectively it should be a set of rules that the component must adhere to in order to be accessible. Something like:

```
The component must:

- communicate number value and label (if present) in a single dictation when read with a screen reader
- convey the meaning of the number shown
```

File: app/assets/stylesheets/govuk_publishing_components/components/_big-number.scss [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+.gem-c-big-number__value {
+  @include govuk-font($size: 80, $weight: bold);
+}
+
+.gem-c-big-number__label {
+  @include govuk-font($size: 16, $weight: bold, $line-height: 2);
+
+  &:before {
+    content: "";
+    display: block;
```
Comment (andysellick) [inline app/assets/stylesheets/govuk_publishing_components/components/_big-number.scss:14]: The reason for this pseudo element is documented nicely in the YML file but I wonder if it would be worth putting a short comment in here as well? The reason being that if a developer might not think to consult the YML file before removing this bit (if they thought it wasn't useful).

File: app/views/govuk_publishing_components/components/docs/big_number.yml [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+shared_accessibility_criteria:
+- link
+examples:
+  default:
+    data:
+      number: 119
+  with_labels:
+    description: |
+      Labels for numbers are given inline styling but stacked using pseudo elements with display block. This is to bypass an issue with NVDA where block level elements are dictated separately.
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/docs/big_number.yml:17]: Nitpick: this description (and a few below) don't need to be formatted using `|` as it's a single line. You could just write `description: Labels for numbers are given...`

File: app/views/govuk_publishing_components/components/_big_number.html.erb [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+  <% unless label.nil? %>
+    <% # add a virtual space here to handle screen readers printing dictations without a space between the number and the label %>
+    <span class="govuk-visually-hidden">&nbsp;</span>
+    <span class="gem-c-big-number__label">
+      <%= label %>
+    </span>
+  <% end %>
+<% end %>
+<% if number %>
```
Comment (andysellick) [inline app/views/govuk_publishing_components/components/_big_number.html.erb:26]: This checks whether or not to render the component, correct? Would it make more sense for it to wrap the whole of the code (i.e. from line 13 onwards) rather than just this part?

File: spec/components/big_number_spec.rb [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+  it "renders nothing when nothing is passed to it" do
+    assert_empty render_component({})
+    assert_empty render_component(number: false)
+  end
+
+  it "renders a number value" do
+    render_component(number: 500)
+
+    assert_select ".gem-c-big-number", count: 1
+    assert_select ".govuk-list", count: 0
```
Comment (andysellick) [inline spec/components/big_number_spec.rb:17]: What's this check for? I don't see the `govuk-list` class used anywhere in the component.

**Review by andysellick — COMMENTED** [review]
Nice. I like the changes you've made - makes the component much clearer and more straightforward.

Just found a few things that need a look, but otherwise looks good 👍

File: app/views/govuk_publishing_components/components/docs/big_number.yml (hunk 12 above)
Comment (owenatgov) [reply app/views/govuk_publishing_components/components/docs/big_number.yml:8]: Your suggestion makes sense to me, updating...

File: spec/components/big_number_spec.rb (hunk 16 above)
Comment (owenatgov) [reply spec/components/big_number_spec.rb:17]: Oops, legacy test. Removing...

**Review by andysellick — APPROVED** [review]
Good stuff. I'm also still particularly enjoying the name of this component.
